### PR TITLE
pmi: refactor the message layer

### PR DESCRIPTION
**Pull request by hzhou**
## Pull Request Description
This is a 1st split from #6168 since that PR is getting too many commits.

Add PMI message layer abstraction.

[skip warnings]

## Author Checklist
* [x] **Provide Description** 
      Particularly focus on _why_, not _what_. Reference background, issues, test failures, xfail entries, etc.
* [x] **Commits Follow Good Practice**
      Commits are self-contained and do not do two things at once. 
      Commit message is of the form: `module: short description` 
      Commit message explains what's in the commit.
* [ ] **Passes All Tests**
      Whitespace checker. Warnings test. Additional tests via comments.
* [x] **Contribution Agreement**
      For non-Argonne authors, check [contribution agreement](http://www.mpich.org/documentation/contributor-docs/). 
      If necessary, request an explicit comment from your companies PR approval manager.

### CODE REVIEW

**Comment by hzhou** [comment]
test:mpich/pmi

**Comment by hzhou** [comment]
test:mpich/pmi

**Comment by hzhou** [comment]
test:mpich/pmi

File: src/pmi/src/pmi_wire.c [hunk 1]
```diff
@@ -14,7 +14,7 @@
 #define IS_SPACE(c) ((c) == ' ')
 #define IS_EOS(c) ((c) == '\0')
 #define IS_EOL(c) ((c) == '\n' || (c) == '\0')
-#define IS_KEY(c) (isalnum(c) || (c) == '_' || (c) == '-' || (c) == '/')
+#define IS_KEY(c) (!(IS_SPACE(c) || IS_EOL(c) || (c) == '='))
```
Comment (raffenet) [inline src/pmi/src/pmi_wire.c:18]: Will users get an error if a put or get is attempted with a bad character?

File: src/pmi/src/pmi_v1.c [hunk 2]
```diff
@@ -886,8 +849,7 @@ static int PMII_singinit(void)
     char port_c[8];
     unsigned short port;
 
-    struct PMIU_cmd pmicmd;
-    PMIU_cmd_init(&pmicmd, USE_WIRE_VER, NULL);
+    struct PMIU_cmd pmicmd = { 0 };
```
[the remark this answers is not in the document]
Comment (hzhou) [reply src/pmi/src/pmi_v1.c:846]: Good catch! Will fix.

File: src/pmi/src/pmi_wire.c (hunk 1 above)
Comment (hzhou) [reply src/pmi/src/pmi_wire.c:18]: It likely will get a parse error, which will become a PMI error and then an MPI error.

**Review by raffenet — APPROVED** [review]
I think the utilities are quite useful as implemented. Long-term, I think we should look at revising the naming. The idea of constructing "messages" seems like something that should be internal to the PMIU_cmd utils and not exposed. To me, something like:
```
pmi_cmd_init
pmi_cmd_set_inputs
pmi_cmd_execute
pmi_cmd_get_outputs
pmi_cmd_free
```
is what we've now achieved. But perhaps my understanding of the workflow isn't quite right.

For context, I'm imagining an implementation where `PMI_KVS_Get` is implemented without the need for communicating with the PMI proxy/server. Maybe the command utils are bypassed entirely in such a scenario and my point doesn't apply.

File: src/pmi/src/pmi_v1.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (hzhou) [reply src/pmi/src/pmi_v1.c]: fixed

**Comment by hzhou** [comment]
test:mpich/ch3/most
test:mpich/ch4/most

**Pull request merged**